### PR TITLE
Fix wrong result due to column position mismatch in BigQuery

### LINKED ISSUE #8183: Trino BigQuery Connector: CAST causing SELECT to return NULL values and inter-changing column positions
I am using **Starburst Trino 356-e (LTS)** on a single node cluster (same node running coordinator and worker).

I created a table from Trino CLI:
<pre>
trino:frg96_dataset> CREATE TABLE IF NOT EXISTS
                  -> test (
                  ->     colvarchar VARCHAR,
                  ->     colsmallint SMALLINT,
                  ->     colint INTEGER,
                  ->     colbigint BIGINT,
                  ->     coldouble DOUBLE,
                  ->     coldecimal DECIMAL(10,3),
                  ->     coldate DATE,
                  ->     coltimestamp TIMESTAMP(3),
                  ->     colvarchar2 VARCHAR,
                  ->     colvarchar3 VARCHAR
                  -> );
CREATE TABLE
</pre>

This created a bigquery table as shown below:

Describe table from Trino:
<pre>
trino:frg96_dataset> describe test;
    Column    |     Type      | Extra | Comment
--------------+---------------+-------+---------
 colvarchar   | varchar       |       |
 colsmallint  | bigint        |       |
 colint       | bigint        |       |
 colbigint    | bigint        |       |
 coldouble    | double        |       |
 coldecimal   | decimal(38,9) |       |
 coldate      | date          |       |
 coltimestamp | timestamp(3)  |       |
 colvarchar2  | varchar       |       |
 colvarchar3  | varchar       |       |
(10 rows)

Query 20210603_042500_00006_59a9y, FINISHED, 1 node
Splits: 19 total, 19 done (100.00%)
1.23 [10 rows, 730B] [8 rows/s, 592B/s]
</pre>

Inserted 5 rows into test via BigQuery Cloud Console:
<pre>
INSERT INTO `gcp-project-id.frg96_dataset.test`
VALUES(
    'Hello world',
    12345,
    1234567,
    123456789,
    12345.678,
    NUMERIC '12345.678',
    DATE '2014-09-17',
    DATETIME '2014-09-17 18:35:54',
    'HELLO WORLD 2',
    'Hello WORLD 3'
);
</pre>
</br>
When I run a select on all columns, The output is correct as shown:
<pre>
trino:frg96_dataset> SELECT
                  ->     colvarchar,
                  ->     colsmallint,
                  ->     colint,
                  ->     colbigint,
                  ->     coldouble,
                  ->     colvarchar,
                  ->     coldecimal,
                  ->     coldate,
                  ->     coltimestamp,
                  ->     colvarchar2,
                  ->     colvarchar3
                  -> FROM test;
 colvarchar  | colsmallint | colint  | colbigint | coldouble | colvarchar  |   coldecimal    |  coldate   |      coltimestamp       |  colvarchar2  |  colvarchar3
-------------+-------------+---------+-----------+-----------+-------------+-----------------+------------+-------------------------+---------------+---------------
 Hello world |       12345 | 1234567 | 123456789 | 12345.678 | Hello world | 12345.678000000 | 2014-09-17 | 2014-09-17 18:35:54.000 | HELLO WORLD 2 | Hello WORLD 3
 Hello world |       12345 | 1234567 | 123456789 | 12345.678 | Hello world | 12345.678000000 | 2014-09-17 | 2014-09-17 18:35:54.000 | HELLO WORLD 2 | Hello WORLD 3
 Hello world |       12345 | 1234567 | 123456789 | 12345.678 | Hello world | 12345.678000000 | 2014-09-17 | 2014-09-17 18:35:54.000 | HELLO WORLD 2 | Hello WORLD 3
 Hello world |       12345 | 1234567 | 123456789 | 12345.678 | Hello world | 12345.678000000 | 2014-09-17 | 2014-09-17 18:35:54.000 | HELLO WORLD 2 | Hello WORLD 3
 Hello world |       12345 | 1234567 | 123456789 | 12345.678 | Hello world | 12345.678000000 | 2014-09-17 | 2014-09-17 18:35:54.000 | HELLO WORLD 2 | Hello WORLD 3
(5 rows)

Query 20210603_042330_00005_59a9y, FINISHED, 1 node
Splits: 17 total, 17 done (100.00%)
4.32 [5 rows, 545B] [1 rows/s, 126B/s]
</pre>
</br>
But When I use CAST for the SMALLINT and INT columns, I get the following as shown below:
<pre>
trino:frg96_dataset>
                  -> SELECT
                  ->     colvarchar,
                  ->     CAST(colsmallint AS SMALLINT) AS colsmallint,
                  ->     CAST(colint AS INTEGER) AS colint,
                  ->     colbigint,
                  ->     coldouble,
                  ->     colvarchar,
                  ->     coldecimal,
                  ->     coldate,
                  ->     coltimestamp,
                  ->     colvarchar2,
                  ->     colvarchar3
                  -> FROM test;
 colvarchar | colsmallint | colint  | colbigint | coldouble | colvarchar | coldecimal |  coldate   | coltimestamp |  colvarchar2  |     colvarchar3
------------+-------------+---------+-----------+-----------+------------+------------+------------+--------------+---------------+---------------------
 NULL       |        NULL | 1234567 | 123456789 |      NULL | NULL       |       NULL | 2014-09-17 | NULL         | HELLO WORLD 2 | 2014-09-17T18:35:54
 NULL       |        NULL | 1234567 | 123456789 |      NULL | NULL       |       NULL | 2014-09-17 | NULL         | HELLO WORLD 2 | 2014-09-17T18:35:54
 NULL       |        NULL | 1234567 | 123456789 |      NULL | NULL       |       NULL | 2014-09-17 | NULL         | HELLO WORLD 2 | 2014-09-17T18:35:54
 NULL       |        NULL | 1234567 | 123456789 |      NULL | NULL       |       NULL | 2014-09-17 | NULL         | HELLO WORLD 2 | 2014-09-17T18:35:54
 NULL       |        NULL | 1234567 | 123456789 |      NULL | NULL       |       NULL | 2014-09-17 | NULL         | HELLO WORLD 2 | 2014-09-17T18:35:54
(5 rows)

Query 20210603_042802_00008_59a9y, FINISHED, 1 node
Splits: 17 total, 17 done (100.00%)
2.15 [5 rows, 545B] [2 rows/s, 253B/s]
</pre>
Certain columns return NULL and certain column positions are interchanged.
</br>
Weirdly, this happens only with this combination of columns. When I remove the last column `colvarchar3` from SELECT, it works fine:
<pre>
trino:frg96_dataset> SELECT
                  ->     colvarchar,
                  ->     CAST(colsmallint AS SMALLINT) AS colsmallint,
                  ->     CAST(colint AS INTEGER) AS colint,
                  ->     colbigint,
                  ->     coldouble,
                  ->     colvarchar,
                  ->     coldecimal,
                  ->     coldate,
                  ->     coltimestamp,
                  ->     colvarchar2
                  -> FROM test;
 colvarchar  | colsmallint | colint  | colbigint | coldouble | colvarchar  |   coldecimal    |  coldate   |      coltimestamp       |  colvarchar2
-------------+-------------+---------+-----------+-----------+-------------+-----------------+------------+-------------------------+---------------
 Hello world |       12345 | 1234567 | 123456789 | 12345.678 | Hello world | 12345.678000000 | 2014-09-17 | 2014-09-17 18:35:54.000 | HELLO WORLD 2
 Hello world |       12345 | 1234567 | 123456789 | 12345.678 | Hello world | 12345.678000000 | 2014-09-17 | 2014-09-17 18:35:54.000 | HELLO WORLD 2
 Hello world |       12345 | 1234567 | 123456789 | 12345.678 | Hello world | 12345.678000000 | 2014-09-17 | 2014-09-17 18:35:54.000 | HELLO WORLD 2
 Hello world |       12345 | 1234567 | 123456789 | 12345.678 | Hello world | 12345.678000000 | 2014-09-17 | 2014-09-17 18:35:54.000 | HELLO WORLD 2
 Hello world |       12345 | 1234567 | 123456789 | 12345.678 | Hello world | 12345.678000000 | 2014-09-17 | 2014-09-17 18:35:54.000 | HELLO WORLD 2
(5 rows)

Query 20210603_044921_00011_59a9y, FINISHED, 1 node
Splits: 17 total, 17 done (100.00%)
2.22 [5 rows, 470B] [2 rows/s, 212B/s]
</pre>

</br>
Any idea(s) what could be causing this issue?

### PR DESCRIPTION
**Pull request by ebyhr**
Fixes #8183

### CODE REVIEW

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryIntegrationSmokeTest.java [hunk 1]
```diff
@@ -317,6 +317,25 @@ public void testRepeatCountAggregationView()
...
+    public void testColumnPositionMismatch()
+    {
+        String tableName = "test.test_column_position_mismatch";
+
+        assertUpdate("DROP TABLE IF EXISTS " + tableName);
+        assertUpdate(format("CREATE TABLE %s (c_varchar VARCHAR, c_int INT, c_date DATE)", tableName));
+        onBigQuery(format("INSERT INTO %s VALUES ('a', 1, '2021-01-01')", tableName));
+
+        assertQuery(
+                "SELECT c_varchar, CAST(c_int AS SMALLINT), c_date FROM " + tableName,
```
Comment (hashhar) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryIntegrationSmokeTest.java:333]: nit: A comment like `// Adding a CAST makes BigQuery return columns in a different order` would help understand the test. It's unclear without referring to the linked issue about how a similar-looking query can cause issues.

**Review by wendigo — APPROVED** [review]

**Review by hashhar — APPROVED** [review]
Thanks for the fix.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryIntegrationSmokeTest.java [hunk 2]
```diff
@@ -317,6 +317,26 @@ public void testRepeatCountAggregationView()
...
+    @Test
+    public void testColumnPositionMismatch()
+    {
+        String tableName = "test.test_column_position_mismatch";
+
+        assertUpdate("DROP TABLE IF EXISTS " + tableName);
+        assertUpdate(format("CREATE TABLE %s (c_varchar VARCHAR, c_int INT, c_date DATE)", tableName));
+        onBigQuery(format("INSERT INTO %s VALUES ('a', 1, '2021-01-01')", tableName));
+
+        // Adding a CAST makes BigQuery return columns in a different order
```
Comment (findepi) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryIntegrationSmokeTest.java:332]: The CAST is not pushed down into the connector (it's not "visible" to the connector).
What was triggering the original problem?

(i have some doubts about https://github.com/trinodb/trino/issues/8170#issuecomment-855324032)

Comment (ebyhr) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryIntegrationSmokeTest.java:332]: @findepi To be more precise, `GenericRecord record` is returned in the fixed order `c_varchar VARCHAR, c_int INT, c_date DATE` regardless of CAST, but `List<BigQueryColumnHandle> columns` order differ if CAST exists. 

Without CAST (varchar -> int -> date)
Screen Shot 2021-06-07 at 18 46 10

With CAST (varchar -> date -> int)
Screen Shot 2021-06-07 at 18 45 28

Do you have new suggested comment? How about `// Adding a CAST changes the column order in BigQuery page source`?

Comment (findepi) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryIntegrationSmokeTest.java:332]: > Adding a CAST changes the column order in BigQuery page source

This is true currently, but may not be true tomorrow.

This is related to https://github.com/trinodb/trino/pull/7756 cc @losipiuk .
I don't think we can reliably test that. **IF** we can, we should test this for all connectors.

Anyway, i am sure CAST was not a necessary condition to trigger the bug, so we should update the release notes entry (eg make more vague).

Comment (ebyhr) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryIntegrationSmokeTest.java:332]: @findepi  Updated as below. Please feel free to edit my comment. 
`Fix incorrect results when accessing BigQuery records with wrong index.`

**Pull request merged**